src/utils/document.js

```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Generate a javascript object reflecting the updates from the document.
+ *
+ * @param {Document} doc - The hadron document.
+ *
+ * @returns {Object} The javascript update object.
+**/
+export const getSetUpdateForDocumentChanges = (doc) => {
+  const object = {};
+
+  if (doc && doc.elements) {
+    for (const element of doc.elements) {
+      if (
+        !element.isRemoved()
+        && element.currentKey !== ''
+        && element.isModified()
+      ) {
+        // Include the full modified element. This way if a nested field
+        // has been altered (changed/added/removed) it is set at the top
+        // most level and exists in the set update document.
+        // We don't dive into nested objects because we can't gaurantee a
+        // path to the element using '.' dot notation will update
+        // the correct field, because field names can now contain dots.
+        object[element.currentKey] = element.generateObject();
```

Comment:
I'll update that comment to make it a bit clearer. What happens is that when a nested field is edited:
```
{
  a: {
    b: 1
  }
}
```
to
```
{
  a: {
    b: 3
  }
}
```
We're running an update on the `a` field, replacing its inner document. The implications of this are that if a field was added to `a` in the background, it is overwritten. That's different from at the top level we aren't overriding new fields because we're updating now.